pallets/gear/src/lib.rs

```diff
@@ -606,6 +606,8 @@ pub mod pallet {
             let code_and_id = CodeAndId::new(code);
             let code_info = CodeInfo::from_code_and_id(&code_and_id);
 
+            let program_id = ProgramId::generate(code_and_id.code_id(), &salt);
```

Comment:
its' wrong. You create a message with another program id below (you call `from_packet` which uses not only salt and code_id, but message id)